src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1599,7 +1676,78 @@ Node TheoryStringsRewriter::rewriteContains( Node node ) {
     Node ret = NodeManager::currentNM()->mkConst(false);
     return returnRewrite(node, ret, "ctn-len-ineq");
   }
-  else if (checkEntailArithEq(len_n1, len_n2))
+  
+  // multi-set reasoning
+  //   For example, contains( str.++( x, "b" ), str.++( "a", x ) ) ---> false
+  //   since the number of a's in the second argument is greater than the number
+  //   of a's in the first argument
+  std::map< Node, unsigned > num_nconst[2];
+  std::map< Node, unsigned > num_const[2];
+  for( unsigned j=0; j<2; j++ ){
+    std::vector< Node >& ncj = j==0 ? nc1 : nc2;
+    for( const Node& cc : ncj ){
+      if( cc.isConst() ){
+        num_const[j][cc]++;
+      }else{
+        num_nconst[j][cc]++;
+      }
+    }
+  }
+  bool ms_success = true;
+  for( std::pair< const Node, unsigned >& nncp : num_nconst[0] )
+  {
+    if( nncp.second>num_nconst[1][nncp.first] ){
+      ms_success = false;
+      break;
+    }
+  }
+  if( ms_success ){
+    // count the number of constant characters in the first argument
+    std::map< Node, unsigned > count_const[2];
+    std::vector< Node > chars;
```

Comment:
With the current rewrite, there doesn't really seem to be a reason to construct the actual `Node`. We could simply keep the characters as `unsigned`s, which would simplify the code and be more efficient.

Reply:
The reason I'm using Node is actually for future work.  See the comment:

// TODO (#1180): count the number of 2,3,4,.. character substrings

The point is that we can generalize this kind of reasoning to multi-character substrings, and for that we will need Node.